### ISSUE TITLE
Mempool never cleared of invalid transaction, so miner can never mine a block

### ISSUE THREAD

**Issue by stringhandler** [tari-project/tari#2101]
**Describe the bug**

A miner asks a base node for a block, but the base node has an invalid transaction in the mempool, for example a double spend or an already mined transaction. Any blocks that the miner mine will be invalid and whenever the miner asks the base node for a block, it includes the transaction.
It's not clear how the mempool gets the invalid transaction, but it could possibly be that it syncs from another base node that is on an older tip. 

The following logs are present (see attached):
```
2020-07-30 09:28:15.275142524 [c::val::block_validators] WARN  Block validation failed because the block has invalid input: bcd8c25773c0c28d2ad63cdf288b9b1882f2a8fc719d5c8bdd7a22fad7ca1338 [OutputFeatures { flags: (empty), maturity: 0 }]

2020-07-30 09:28:15.275156299 [c::cs::database] WARN  Orphan block 65517 (d7ee2885566cd7891fca45a30e26eb86d695cec27d23684499657374fb9304f7) failed validation during chain reorg.
2020-07-30 09:28:15.281637759 [c::cs::data
```

[mining-logs.txt.zip](https://github.com/tari-project/tari/files/5007246/mining-logs.txt.zip)

**Comment by stringhandler** [comment]
I have a feeling something like this is happening:
```
1. Node 1 has Tx with excess sig xs_a in its mempool
2. Node 2 starts up, and starts syncing and gets xs_a in its mempool
3. Node 1 receives a block and removes xs_a from the mempool
4. Node 2 eventually syncs, but somehow still has xs_a
5. Node 3 starts, and receives xs_a from Node 2
6. Node 2 and 3 continue to receive blocks, but will never remove xs_a because none of the blocks it receives will contain xs_a
7. In addition, any nodes that sync with Node 2 and 3 will receive xs_a
```

Could be wrong though

**Comment by SWvheerden** [comment]
Thats sound very likely, I agree with this 100%. 

We either have to clear/check the mempool after sync is done, or we have check each transaction as we put in in the mempool

**Comment by sdbondi** [comment]
Sounds very plausible - a call to `txn.validate_internal_consistency` in the mempool sync protocol before inserting would fix it.

EDIT: Actually the `Mempool::insert` function already runs the validators. So we do check each transaction as they are inserted.

**Comment by delta1** [comment]
Firstly my apologies if I'm asking obvious stuff :) 

Going through that log @mikethetike at the first signs of the issue: 

|log line no|  event/s |
|----|----------|
|32637|chain synced to #51502 (72816f)|
|32645| miner mines candidate block #51503 (**20fa2a**) |
|32699| candidate block fails validation and is orphaned due to transaction *306a4e* |
|- | blocks #51504 (841f8e) and #51503 (**f38510**) are received and validated from the network |
|32724 | blocks #51504 (841f8e) and #51503 (**f38510**) are added to the chain|
|- |blocks #51505 - #51508 are received, validated and added to the chain |
|32797 | miner mines candidate block #51509 (90145a) |
|32883 | candidate block fails validation and is orphaned due to transaction *306a4e*| 

So - as suspected - a transaction (*306a4e*) was not removed from mempool after receiving new blocks.. 

From what I can see in the code:

[MempoolStorage::process_published_block](/tari-project/tari/blob/development/base_layer/core/src/mempool/mempool_storage.rs#L119) is called to update the mempool after receiving the published block/s - in turn calling [UnconfirmedPool::remove_published_and_discard_double_spends](https://github.com/tari-project/tari/blob/development/base_layer/core/src/mempool/unconfirmed_pool/unconfirmed_pool.rs#L200) 

And that's where I could use a hand figuring out if that's the source of this issue? :) 

There is a [passing test](https://github.com/tari-project/tari/blob/development/base_layer/core/src/mempool/unconfirmed_pool/unconfirmed_pool.rs#L327) for that function in the same file, so it seems to be working correctly...

Is there a logic bug higher up? process_published_block is inserting the published transactions from the unconfirmed pool into the reorg_pool - where is the miner getting the transactions from? 

Any assistance would be appreciated, thanks @sdbondi and @SWvheerden for your insights too

**Comment by SWvheerden** [comment]
I have not seen the logs, but does the logs show where tx 306a4e comes from?

Somewhere the mempool is mising a cleanup or the mempool is accepting a wrong tx's. We need to see where that tx comes from to know where to look

**Comment by SWvheerden** [comment]
But to answer your question. The miner gets the tx's from the mempool. The mempool publishes an event after it processes a block. That even triggers the miner to ask for a new block. 

You can take a look at the miner code, and or the inbound handlers. The miner uses get block template to get a list of transactions from the mempool

**Comment by delta1** [comment]
After the discussions with @SWvheerden (thanks!) we believe this specific issue occurs in blockchain_database.rs at [handle_reorg](https://github.com/tari-project/tari/blob/development/base_layer/core/src/chain_storage/blockchain_database.rs#L1585) 

In the linked log file above, 2 blocks are added to the chain 51501 and 51502, but because no blocks were removed, the function returns `BlockAddResult::Ok` which does not trigger the same mempool check and update that `BlockAddResult::ChainReorg` triggers. 

The proposed fix is to also return `BlockAddResult::ChainReorg` if `removed_blocks` is empty **and** `added_blocks` has any length greater than 1 - ensuring that the reorg mempool logic is triggered. 

~Given that there are no end-to-end tests including these functions (please correct me if I'm wrong!) - I am going to factor out this return result into its own function and add some unit tests to that.~ 
Update: Adding a test to `chain_storage_tests` for this

Any other suggestions or comments are welcome

**Comment by delta1** [comment]
There's a further complication.. 

`thread 'tokio-runtime-worker' panicked at 'Added empty set of blocks on reorg.', tari/base_layer/core/src/mempool/mempool_storage.rs:169:31`

By triggering the reorg logic when no blocks are removed, and more than one are added - `Mempool::process_reorg` panics at 

```rust
 let prev_tip_height = removed_blocks
     .last()
     .expect("Added empty set of blocks on reorg.")
     .header
     .height;
```

Since `prev_tip_height` is only used to check for time-locked txs in the unconfirmed pool to be removed, would it suffice that if no blocks are being removed then this check does not need to be done? 

If so, I'll update `Mempool::process_reorg` to first check if `removed_blocks` is not empty, before doing the check for time-locked txs in the unconfirmed pool.

**Comment by stringhandler** [comment]
Closed in #2166
